iroh-embed/src/lib.rs

```diff
@@ -0,0 +1,240 @@
+
+/// An address allowing internal communication with the iroh p2p service.
+///
+/// This is only needed to wire up the several services together and should not be used to
+/// communicate directly to the service.
+///
+/// Can be created from [`p2p::P2pService::addr`].
+pub use iroh_rpc_types::store::StoreClientAddr;
+
+mod p2p;
+mod store;
+
+pub use p2p::P2pService;
+pub use store::RocksStoreService;
+
+/// Builder for an [`Iroh`] system.
+///
+/// At least a store and p2p service must be added using the [`IrohBuilder::with_store`] and
+/// [`IrohBuilder::with_p2p`].
+///
+/// # Examples
+///
+/// ```no_run
+/// use iroh_embed::{Iroh, IrohBuilder, Libp2pConfig, P2pService, RocksStoreService};
+/// use testdir::testdir;
+/// # tokio_test::block_on(async {
+/// let dir = testdir!();
+/// let store = RocksStoreService::new(dir.join("store")).await.unwrap();
+/// let mut p2p_config = Libp2pConfig::default();
+/// p2p_config.listening_multiaddrs = vec![
+///     "/ip4/0.0.0.0/tcp/0".parse().unwrap(),  // random port
+///     "/ip4/0.0.0.0/udp/0/quic-v1".parse().unwrap(),  // random port
+/// ];
+/// let p2p = P2pService::new(p2p_config, dir, store.addr()).await.unwrap();
+/// let _iroh: Iroh = IrohBuilder::new()
+///                     .with_store(store)
+///                     .with_p2p(p2p)
+///                     .build()
+///                     .await
+///                     .unwrap();
+/// # })
+/// ```
+#[derive(Debug)]
+pub struct IrohBuilder {
+    store: Option<RocksStoreService>,
+    p2p: Option<P2pService>,
+    http_resolvers: Vec<String>,
+    indexer: Option<IndexerUrl>,
+}
+
+impl Default for IrohBuilder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl IrohBuilder {
+    /// Creates a new [`IrohBuilder`].
+    pub fn new() -> Self {
+        Self {
+            store: None,
+            p2p: None,
+            http_resolvers: vec![],
+            indexer: Some(IndexerUrl::default()),
+        }
+    }
+
+    /// Adds a store service.
+    ///
+    /// Every [`Iroh`] system needs a store so this can not be skipped.
+    pub fn with_store(mut self, store: RocksStoreService) -> Self {
+        self.store = Some(store);
+        self
+    }
+
+    /// Adds a p2p service.
+    ///
+    /// This service **must** have been built using the address from the store service
+    /// passed to [`IrohBuilder::with_store`].
+    ///
+    /// Every [`Iroh`] system currently needs a p2p service so this can not be skipped.
+    pub fn with_p2p(mut self, p2p: P2pService) -> Self {
+        self.p2p = Some(p2p);
+        self
+    }
+
+    /// Adds IPFS HTTP gateways as resolvers to the iroh system.
+    ///
+    /// IPFS HTTP gateways can be used to resolve IPFS content in addition to retrieving the
+    /// content from the IPFS peer-to-peer system.
+    // TODO: Allow using this without a p2p node?
+    pub fn with_http_resolvers(mut self, http_resolvers: impl Iterator<Item = Url>) -> Self {
```

Comment:
nit: There is different naming conventions, but I associate `with_` always with an alternative `new` constructor, rather than an additive method.